src/eclipse.rs

```diff
@@ -0,0 +1,22 @@
+use std::result::Result;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Vote {
+    signature: Vec<u8>,
+    public_key: Vec<u8>,
+    message: Vec<u8>,
+}
+
+impl Vote {
+    pub fn new(signature: Vec<u8>, public_key: Vec<u8>, message: Vec<u8>) -> Self {
+        Vote {
+            signature,
+            public_key,
+            message,
+        }
+    }
+}
+
+pub trait ProofGenerator {
+    fn generate_proof(&self, slot: u64, votes: Vec<Vote>) -> Result<(), String>;
```

Comment:
Chime in if you have any thoughts about this.

Reply:
Just an observation: I am not sure what the ALEO tx builder will take for the data payload, so at some point we will probably want to return the correct type.

Reply:
> at some point we will probably want to return the correct type.

Could be. There might also be some intermediate component here, that would send the resulting TX to Aleo blockchain, because returning it all the way back to Solana client doesn't make so much sense, but yeah, this is still in the beginning :slightly_smiling_face:

Reply:
Wait.. just to be clear, the result is to be added to Aleo TX here right? 

To submit to Solana, we will be submitting the Aleo block proof, which we will get from the Aleo chain itself.

Reply:
> Wait.. just to be clear, the result is to be added to Aleo TX here right?

That was my thinking. We gather signatures for a given Solana Slot, we generate a proof of that and then submit an Aleo transaction (to Aleo chain).

> To submit to Solana, we will be submitting the Aleo block proof, which we will get from the Aleo chain itself.

Frankly, I haven't thought about this too much yet, but wouldn't this go so that when our custom transaction (with its proof) is in Aleo blockchain, then this proof could be taken and verified as valid in Solana program (a smart contract implementing required operations, once the backing infrastructure i.e. opcodes & such are implemented) ?